tests/smoke/gmres.cpp

```diff
@@ -0,0 +1,893 @@
+        return std::complex< BaseScalarType >( re, im );
+}
+
+/**
+ * Generate random lienar problem matrix data (A) and
+ * the corresponding preconditioner matrix data (P).
+ */
+template< typename NonzeroType, typename DimensionType >
+void generate_random_data(
+	const DimensionType &n,
+	const DimensionType &nz_per_row,
+	std::vector< DimensionType > &MatAveci,
+	std::vector< DimensionType > &MatAvecj,
+	std::vector< NonzeroType > &MatAvecv,
+	std::vector< DimensionType > &MatPveci,
+	std::vector< DimensionType > &MatPvecj,
+	std::vector< NonzeroType > &MatPvecv
+) {
+	Semiring<
+		grb::operators::add< NonzeroType >, grb::operators::mul< NonzeroType >,
+		grb::identities::zero, grb::identities::one
+	> ring;
+	operators::divide< NonzeroType > divide;
+	const NonzeroType one = ring.template getOne< NonzeroType >();
+
+	MatAveci.resize( n * nz_per_row );
+	MatAvecj.resize( n * nz_per_row );
+	MatAvecv.resize( n * nz_per_row );
+	MatPveci.resize( n );
+	MatPvecj.resize( n );
+	MatPvecv.resize( n );
+
+	DimensionType acout = 0;
+	DimensionType pcout = 0;
+	for( DimensionType i = 0; i < n; ++i ){
+		for(
+			DimensionType j = i - std::min( nz_per_row / 2, i );
+			j < std::min( i + nz_per_row / 2, n );
+			++j
+		) {
+			NonzeroType tmp = random_value< NonzeroType >();
+			NonzeroType tmp2 = random_value< NonzeroType >();
+			// make sure preconditioner is useful
+			if( i == j ) {
+				tmp2 = tmp2 * static_cast< NonzeroType >( n * n );
+				tmp = tmp + tmp2;
+			}
+			MatAveci[ acout ] = i;
+			MatAvecj[ acout ] = j;
+			MatAvecv[ acout ] = tmp;
+			++acout;
+			if( i == j ) {
+				MatPveci[ pcout ] = i;
+				MatPvecj[ pcout ] = j;
+				grb::foldr( one, tmp2, divide );
+				MatPvecv[ pcout ] = tmp2;
+				++pcout;
+			}
+		}
+	}
+	MatAveci.resize( acout );
+	MatAvecj.resize( acout );
+	MatAvecv.resize( acout );
+}
+
+/**
+ * Build random matrix A and the preconditioner P.
+ */
+RC make_matrices(
+	Matrix< ScalarType > &A,
+	Matrix< ScalarType > &P,
+	const size_t &n,
+	const size_t &nz_per_row
+){
+	RC rc = SUCCESS;
+
+	{
+		std::vector< size_t > MatAveci;
+		std::vector< size_t > MatAvecj;
+		std::vector< ScalarType > MatAvecv;
+		std::vector< size_t > MatPveci;
+		std::vector< size_t > MatPvecj;
+		std::vector< ScalarType > MatPvecv;
+		std::srand( 0 );
+		generate_random_data(
+			n,
+			nz_per_row,
+			MatAveci,
+			MatAvecj,
+			MatAvecv,
+			MatPveci,
+			MatPvecj,
+			MatPvecv
+		);
+		rc = rc ? rc : buildMatrixUnique( A, MatAveci.begin(), MatAvecj.begin(), MatAvecv.begin(), MatAveci.size(), SEQUENTIAL );
+		rc = rc ? rc : buildMatrixUnique( P, MatPveci.begin(), MatPvecj.begin(), MatPvecv.begin(), MatPveci.size(), SEQUENTIAL );
+	}
+
+	return rc;
+}
+
+/**
+ * Solves the  least linear square problem defined by vector H[1:n] x =  H[ 0 ],
+ * using Givens rotations and backsubstitution. The results is stored in H[ 0 ],
+ * which is sused to update GMRES solution, vector x.
+ */
+// todo: this is a temp implementations and
```

Comment:
TODOs can go into the doxygen (`\todo Replace by ALP/Dense calls once available`, e.g.)